File: Top200/474-FindMaxForm.go
package main

import "math"

func findMaxForm(strs []string, m int, n int) int {
	strsLen := len(strs)
	counts0 := make([]int, strsLen)
	counts1 := make([]int, strsLen)

	for i := 0; i < strsLen; i++ {
		count := countStr(strs[i])
		counts0[i] = count[0]
		counts1[i] = count[1]
	}

	dp := make([][]int, m+1)
	for i := 0; i < m+1; i++ {
		dp[i] = make([]int, n+1)
	}

	for i := 0; i < strsLen; i++ {
		for j := m; j >= counts0[i]; j-- {
			for k := n; k >= counts1[i]; k-- {
				dp[j][k] = int(math.Max(float64(dp[j][k]), float64(dp[j-counts0[i]][k-counts1[i]]+1)))
			}
		}
	}

	return dp[m][n]
}

func countStr(str string) []int {
	count0 := 0
	count1 := 0
	for _, c := range str {
		if c == '0' {
			count0++
		} else {
			count1++
		}
	}

	return []int{count0, count1}
}

File: Top200/474-FindMaxForm.java
package Top200;

class FindMaxForm {
    public int findMaxForm(String[] strs, int m, int n) {
        int[] zeros = new int[strs.length];
        int[] ones = new int[strs.length];

        for (int i = 0; i < strs.length; i++) {
            int[] count = countStr(strs[i]);
            zeros[i] = count[0];
            ones[i] = count[1];
        }

        int[][] dp = new int[m + 1][n + 1];
        

        for (int i = 0; i < strs.length; i++) {
            for (int j = m; j >= zeros[i]; j--) {   // 这里的顺序需要注意
                for (int k = n; k >= ones[i]; k--) {
                    dp[j][k] = Math.max(dp[j][k], dp[j-zeros[i]][k-ones[i]] + 1);
                }
            }
        }

        return dp[m][n];
    }

    public int[] countStr(String str) {
        if (str == null || str.length() == 0) {
            return new int[]{0, 0};
        }

        int zeroCount = 0;
        int oneCount = 0;
        for (int i = 0; i < str.length(); i++) {
            if (str.charAt(i) == '1') {
                oneCount++;
            } else {
                zeroCount++;
            }
        }

        return new int[]{zeroCount, oneCount};
    }
}